Feature - User Authentication & Authorisation.

mansoor090 opened the pull request.

mansoor090 committed:
Created PUT endpoint to practice CounterMethod.
Created POST endpoint to practice raw JSON body type as PARAMS.
Added user schema using Prisma with Postgres.
DB connected via .env.

mansoor090 committed:
Fixed Prisma Error - Prisma not recognising Prisma Client default path location.
Added Put/Delete endpoints for user delete/update operations.

mansoor090 committed:
Added Authentication using Passport-Local via Guard&Strategy
Added Authorisation using Passport-JWT via Guard&Strategy
Added Password field to User Table

junaidnza commented on src/auth/strategies/jwt.strategy.ts:
```diff
@@ -0,0 +1,21 @@
+    super({
+      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+      ignoreExpiration: false,
+      secretOrKey: 'abc123$',
```
Lets use env variable for this value.

junaidnza commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,22 @@
+}
+
+model User {
+  id           Int      @id @default(autoincrement())
```
Lets use uuid instead of simple int.
Read about what is uuid & what are the benefits of benefits of uuid.

junaidnza commented on src/auth/dto/authPayloadDto.ts:
```diff
@@ -0,0 +1,10 @@
+export class AuthPayloadDto {
+  @IsString()
+  @IsNotEmpty()
+  username: string;
```
Why are we getting username if we do not have any value of username in DB ?

junaidnza replied:
We should be expecting email.
And please also add a validation for email format as well here.

mansoor090 committed:
Fixed Authentication issue in Login endpoint
Fixed Hashing algorithm for secretkey
Connected Database with Auth/Login, User/Create/Update
